### ISSUE TITLE
LDAP Support not working Right

### ISSUE THREAD

**Issue by Alsterdetektive1** [mealie-recipes/mealie#1027]
### First Check

- [X] This is not a feature request
- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.

### What is the issue you are experiencing?

I cant login in Mealie with LDAP. I think it's because anonymous access is deactivated on my LDAP, but I think it's a kind of Bug, i cant specify a login dn.

### Deployment

Docker (Linux)

### Mealie Version

newest

**Comment by achurak** [comment]
I think I have the same issue, though it has nothing to do with anonymous access (at least in my case). Looks like there's something wrong with the part of the code that's supposed to check if the user already exists in the database and create a new one if it doesn't. Basically what it means is that I can sign in only once because after I log out and try to sign back in, mealie keeps trying to insert my user account into the db again thus breaking the unique constraint:
```
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: users.username
[SQL: INSERT INTO users (created_at, update_at, id, full_name, username, email, password, admin, advanced, group_id, cache_key, can_manage, can_invite, can_organize, owned_recipes_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]
[parameters: ('2022-03-10 16:45:38.794491', '2022-03-10 16:45:38.794504', '184e039a429e40279a9044a47466cdb2', 'achurak', 'achurak', 'achurak', 'LDAP', 0, 0, '1d34054d9fd5420781edc03a371a00b9', '1234', 0, 0, 0, None)]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
```

**Comment by pratio** [comment]
@Alsterdetektive1  I think you might be on to something with the anonymous authentication issue.

I have openldap setup and it works fine on my calibre-web installation but not on mealie. When i check the implementation this is what i see

```
>>> import ldap
>>> import os
>>> server_url = os.environ.get('LDAP_SERVER_URL')
>>> user_name='mike'
>>> user_dn = os.environ.get('LDAP_BIND_TEMPLATE').format(user_name)
>>> server_url
'ldap://ldap:1389'
>>> user_dn
'cn=mike,dc=example,dc=org'
>>> conn = ldap.initialize(server_url)
>>> 
>>> conn.simple_bind_s(user_dn, 'mypassword)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/pysetup/.venv/lib/python3.9/site-packages/ldap/ldapobject.py", line 249, in simple_bind_s
    resp_type, resp_data, resp_msgid, resp_ctrls = self.result3(msgid,all=1,timeout=self.timeout)
  File "/opt/pysetup/.venv/lib/python3.9/site-packages/ldap/ldapobject.py", line 543, in result3
    resp_type, resp_data, resp_msgid, decoded_resp_ctrls, retoid, retval = self.result4(
  File "/opt/pysetup/.venv/lib/python3.9/site-packages/ldap/ldapobject.py", line 553, in result4
    ldap_result = self._ldap_call(self._l.result4,msgid,all,timeout,add_ctrls,add_intermediates,add_extop)
  File "/opt/pysetup/.venv/lib/python3.9/site-packages/ldap/ldapobject.py", line 128, in _ldap_call
    result = func(*args,**kwargs)
ldap.INVALID_CREDENTIALS: {'msgtype': 97, 'msgid': 1, 'result': 49, 'desc': 'Invalid credentials', 'ctrls': []}
```
When you ssh into the container, in the root directory you can also try this with
```
>>> from app.mealie.core import security
>>> 
```

The connection to ldap works

```
ldapsearch -x -H ldap://ldap:1389 -D "cn=admin,dc=example,dc=org" -w 'admin' -b "dc=example,dc=org"
```
as i can see from the container itself

```
# user01, users, example.com
dn: cn=user01,ou=users,dc=example,dc=com
cn: User1
cn: user01
sn: Bar1
objectClass: inetOrgPerson
objectClass: posixAccount
objectClass: shadowAccount
userPassword:: cGFzc3dvcmQx
uid: user01
uidNumber: 1000
gidNumber: 1000
homeDirectory: /home/user01

# user02, users, example.com
dn: cn=user02,ou=users,dc=example,dc=com
cn: User2
cn: user02
sn: Bar2
objectClass: inetOrgPerson
objectClass: posixAccount
objectClass: shadowAccount
userPassword:: cGFzc3dvcmQy
uid: user02
uidNumber: 1001
gidNumber: 1001
homeDirectory: /home/user02
```
So I'm sure there's something we can do

**Comment by Alsterdetektive1** [comment]
I was able to get it work with non anymous authentication.  I will write later, what I found out. Now my newer Problem is, that every User can edit and delete every recipe. Even if He is in another group. I didn't get the Beta to work, so sadly I needed to move to another recipe manager.

**Comment by pratio** [comment]
@Alsterdetektive1 Yes, that's true. Fortunately, I've never had a problem with that. 
but if you could shed some light on how you got the auth to work, that'd be great

**Comment by pratio** [comment]
Okay, I figured it out. 
@hay-kot  apologies mate, absolutely nothing wrong with your implementation, it works as it should. 

This is what was wrong with my setup.
The Base DN was incorrect
I was making a mistake in passing the `base_dn` and it took a while to figure out the issue.
This is what helped me figure it out
https://gist.github.com/ibeex/1288159
I tried to use simple auth as shown here and it showed me that the `base_dn` was wrong 


I'm not sure if i should close the issue here, since I didn't open it and it seems that the original poster of the issue also got it work

**Comment by pratio** [comment]
> I think I have the same issue, though it has nothing to do with anonymous access (at least in my case). Looks like there's something wrong with the part of the code that's supposed to check if the user already exists in the database and create a new one if it doesn't. Basically what it means is that I can sign in only once because after I log out and try to sign back in, mealie keeps trying to insert my user account into the db again thus breaking the unique constraint:

This has not been my experience, it's working as it should and does not create the user again. But I'm not doing that with the admin account.

**Comment by Alsterdetektive1** [comment]
So I will tell my false configuration here, so other people will know how to configure it.
The main config is the LDAP_BIND_TEMPLATE.
In other projects you configure their a user and in LDAP_BIND_PASSWORD a password to login. In this project it is more like the LDAP_BASE_DN. You configure where to search for the user and let the username field as {}. Mealie is then loggin in with this information.
For my example it is LDAP_BASE_DN: uid={},ou=group,dc=ldap,dc=mysite,dc=com

Then mealie Set your input username in so you have uid=MyUsername,ou=group,dc=ldap,dc=mysite,dc=com and logging you in with your password. This is working right, but i was a bit confused because of other projects. The problem with the mealie account-rights is still there, but i read that it will be fixed in V1.0, so I think I will try it if it is out.
